substrate/pkg/controller/substrate/cluster/instance.go

```diff
@@ -20,68 +20,109 @@ import (
 	"strings"
 
 	"github.com/aws/aws-sdk-go/aws"
-	"github.com/aws/aws-sdk-go/aws/awserr"
 	"github.com/aws/aws-sdk-go/service/autoscaling"
+	"github.com/aws/aws-sdk-go/service/ec2"
 	"github.com/awslabs/kit/substrate/pkg/apis/v1alpha1"
 	"github.com/awslabs/kit/substrate/pkg/utils/discovery"
 	"knative.dev/pkg/logging"
 	"sigs.k8s.io/controller-runtime/pkg/reconcile"
 )
 
 type Instance struct {
+	EC2         *ec2.EC2
 	AutoScaling *autoscaling.AutoScaling
 }
 
-func (a *Instance) Create(ctx context.Context, substrate *v1alpha1.Substrate) (reconcile.Result, error) {
-	if len(substrate.Status.PublicSubnetIDs) == 0 {
+func (i *Instance) Create(ctx context.Context, substrate *v1alpha1.Substrate) (reconcile.Result, error) {
+	if len(substrate.Status.PublicSubnetIDs) == 0 || substrate.Status.Cluster.LaunchTemplateVersion == nil {
 		return reconcile.Result{Requeue: true}, nil
 	}
-	if _, err := a.AutoScaling.CreateAutoScalingGroupWithContext(ctx, &autoscaling.CreateAutoScalingGroupInput{
-		AutoScalingGroupName: discovery.Name(substrate),
-		DesiredCapacity:      aws.Int64(1), MaxSize: aws.Int64(1), MinSize: aws.Int64(1),
-		LaunchTemplate:    &autoscaling.LaunchTemplateSpecification{LaunchTemplateName: discovery.Name(substrate), Version: aws.String("$Latest")},
-		VPCZoneIdentifier: aws.String(strings.Join(substrate.Status.PublicSubnetIDs, ",")),
-		Tags: []*autoscaling.Tag{{
-			Key:               aws.String(discovery.OwnerTagKey),
-			Value:             aws.String(substrate.Name),
-			PropagateAtLaunch: aws.Bool(true),
-		}, {
-			Key:               aws.String("Name"),
-			Value:             discovery.Name(substrate),
-			PropagateAtLaunch: aws.Bool(true),
-		}},
-	}); err != nil {
-		if strings.Contains(err.Error(), "Invalid IAM Instance Profile name") { // Instance Profile can take several seconds to propagate
+	instancesOutput, err := i.EC2.DescribeInstancesWithContext(ctx, &ec2.DescribeInstancesInput{})
```

Comment:
I think we should either use some filtering here or pagination because if an account/region has a large number of instances this can be an expensive call listing all the instances and it might timeout.

Reply:
Nice catch!!